### PR TITLE
Add entities

### LINKED ISSUE #6: Feature: Item, ItemType, ItemRariy, ItemQuality classes
We need an entity for items, item types, and a valueobject enum for the quality.

# Entity/Item
- ItemId Id
- ItemTypeId TypeId
- ItemType Type
- ItemQuality Quality
- UserId OwnerId
- User Owner

# Entity/ItemType
- ItemTypeId Id
- string Name
- string? Description
- float QualityMin // the minimum quality this item appears in
- float QualityMax // the maximum quality this item can have
- bool StatTrackAvailable // whether or not stat track is available
- string ImageUrl
## methods
// creates a new item, who belongs to the user specified.
// the rarity and the quality are randomly generated.
// only 25% of items are stat track
- Item NewItem(User owner);

# ValueObject/ItemQuality
- float Value
- bool IsStatTrack

### PR DESCRIPTION
**Pull request by gogodze**
adding entities to the project

### CODE REVIEW

File: src/Domain/Entities/Item.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+using Astro.Generated;
+using Domain.Aggregates;
+using Domain.ValueObjects;
+
+namespace Domain.Entities;
+
+public class Item
```
Comment (ddjerqq) [inline src/Domain/Entities/Item.cs]: - this should be sealed
- it should have a `[StrongId(typeof(Ulid))]`
- all entities must inherit from Entity<TId>
  ```cs
  [StrongId(typeof(Ulid))]
  //                       v will get generated when you build the project
  public sealed class Item(ItemId id) : Entity<ItemId>(id)
  // keep in mind that the Id property belongs to the Entity base class
  ```

File: src/Domain/Entities/Item.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+{
+    private ItemId Id { get; set; }
```
Comment (ddjerqq) [inline src/Domain/Entities/Item.cs]: - these should not be private
- also most of these should be immutable (`{ get; init; }`)
- and if we need any of these fields to be mutable, we should encapsulate them and use `{ get; private set; }`

File: src/Domain/Entities/Item.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    private ItemTypeId TypeId { get; set; } 
+    private ItemType Type { get; set; }
+    ItemQuality Quality { get; set; }
+    UserId OwnerId { get; set; }
```
Comment (ddjerqq) [inline src/Domain/Entities/Item.cs]: - also for the warnings, just 
  ```cs
  // to make the compiler stop complaining
  //                                  vvvvvvvvvv
  public UserId OwnerId { get; set; } = default!
  ```

File: src/Domain/Entities/ItemType.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+namespace Domain.Entities;
+
+public class ItemType
```
Comment (ddjerqq) [inline src/Domain/Entities/ItemType.cs]: same stuff goes here about entities

File: src/Domain/ValueObjects/ItemQuality.cs [hunk 5]
```diff
@@ -0,0 +1,7 @@
+namespace Domain.ValueObjects;
+
+public class ItemQuality
```
Comment (ddjerqq) [inline src/Domain/ValueObjects/ItemQuality.cs]: this should be a record - immutable

File: src/Domain/Entities/ItemType.cs [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+{
+    ItemTypeId Id { get; set; } 
+    string Name { get; set; }
+    string? Description { get; set; } 
+    float QualityMin { get; set; } // the minimum quality this item appears in
+    float QualityMax { get; set; } // the maximum quality this item can have
+    bool StatTrackAvailable { get; set; }// whether or not stat track is available
+    string ImageUrl { get; set; }
+}
```
Comment (ddjerqq) [inline src/Domain/Entities/ItemType.cs:36]: add the appropriate methods, according to: [the methods section](https://github.com/ddjerqq/Astro/issues/6#methods)

**Comment by ddjerqq** [comment]
also after you implement the classes, 
add the inventory field to the user [comment](https://github.com/ddjerqq/Astro/issues/6#issuecomment-2274003849)

**Comment by ddjerqq** [comment]
after you add the inventory field to the user, 
configure the entities that you have created, using `IEntityTypeConfiguration<Item/ItemType>` inside of `Persistence/Configurations`
and then create the appropriate migrations

**Comment by ddjerqq** [comment]
and then create controllers to get all of the items, and item types and so on, and that you will be done

**Review by ddjerqq — CHANGES_REQUESTED** [review]

File: src/Domain/Aggregates/User.cs [hunk 7]
```diff
@@ -14,4 +15,6 @@ public sealed class User(UserId id) : AggregateRoot<UserId>(id)
     public string PasswordHash { get; init; } = default!;
     
     public Wallet Wallet { get; set; } = new Wallet(0);
+
+    public ICollection<Item> Inventory {get; set; } = default!;
```
Comment (ddjerqq) [inline src/Domain/Aggregates/User.cs]: alt shift f here for formatting

File: src/Domain/Entities/ItemType.cs [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    public string? Description { get; init; } 
+    public float QualityMin { get; init; } // the minimum quality this item appears in
+    public float QualityMax { get; init; } // the maximum quality this item can have
+    public bool StatTrackAvailable { get; init; }// whether or not stat track is available
+    public string ImageUrl { get; init; }
+    
+    public Item CreateItem(User owner)
+    {
+        var random = Random.Shared.NextSingle() * (QualityMax - QualityMin) + QualityMin;
+        var isStatTrak = Random.Shared.Next(100) < 25;
```
Comment (ddjerqq) [inline src/Domain/Entities/ItemType.cs]: only if `this.IsStatTrackAvailable`

File: src/Domain/Entities/ItemType.cs [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+        return new Item(ItemId.New())
+        {
+            TypeId = Id,
+            Type = this,
+            Quality = new ItemQuality(random, isStatTrak),
+            OwnerId = new UserId(),
```
Comment (ddjerqq) [inline src/Domain/Entities/ItemType.cs]: you need the `owner.Id` here

File: src/Domain/ValueObjects/ItemQuality.cs [hunk 10]
```diff
@@ -0,0 +1,7 @@
+namespace Domain.ValueObjects;
+
+public record ItemQuality(float Value, bool statTrak)
+{
+    float Value { get; set; } = Value;
```
Comment (ddjerqq) [inline src/Domain/ValueObjects/ItemQuality.cs]: records auto generate the init-only properties, so you dont need to type them explicitly

**Review by ddjerqq — CHANGES_REQUESTED** [review]

File: src/Domain/Entities/ItemType.cs [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    public string? Description { get; init; }
+    public float QualityMin { get; init; } // the minimum quality this item appears in
+    public float QualityMax { get; init; } // the maximum quality this item can have
+    public bool StatTrackAvailable { get; init; } // whether or not stat track is available
+    public string? ImageUrl { get; init; }
+
+    public Item CreateItem(User owner)
+    {
+        var random = Random.Shared.NextSingle() * (QualityMax - QualityMin) + QualityMin;
+        bool isStatTrack = false;
```
Comment (ddjerqq) [inline src/Domain/Entities/ItemType.cs:21]: you could simplify this boolean expression like
```cs
var isItemStatTrack = IsStatTrackAvailable && Random <= 25
```

**Review by ddjerqq — APPROVED** [review]

File: src/Domain/Entities/ItemType.cs (hunk 11 above)
Comment (gogodze) [reply src/Domain/Entities/ItemType.cs:21]: should i add this to the code? or its just a feedback?

**Pull request merged**